runtimevar/filevar/filevar.go

```diff
@@ -111,68 +111,58 @@ func (w *watcher) WatchVariable(ctx context.Context) (driver.Variable, error) {
 		// Continue.
 	}
 
-	// Detect if there has been a changed first. If there are no changes, then proceed to add a
-	// fsnotify watcher.
-	wait := true
-	for wait {
-		sd := w.processFile()
-		switch sd.state {
-		case errorState:
-			return zeroVar, sd.err
+	// Start watching over the file and wait for events/errors.
+	// We'll get a notifierErr if the file doesn't currently exist.
+	// Also, note that we may never use the notifier if there's already
+	// a change to the file. We must create it now before checking to
+	// avoid race conditions.
+	notifierErr := w.notifier.Add(w.file)
+	if notifierErr == nil {
+		defer w.notifier.Remove(w.file)
+	}
 
-		case hasChangedState:
-			return *sd.variable, nil
+	for {
+		if v, err := w.processFile(); err != nil {
+			return zeroVar, err
+		} else if v != nil {
+			return *v, nil
+		}
 
-		case stillDeletedState:
-			// Last known state is deleted, need to wait for file to show up before adding a watch.
+		if w.isDeleted {
+			// Last known state is deleted, need to wait for file to show up.
 			t := time.NewTimer(w.waitTime)
 			select {
 			case <-t.C:
 			case <-ctx.Done():
 				t.Stop()
 				return zeroVar, ctx.Err()
 			}
-
-		case noChangeState:
-			wait = false
+			continue
 		}
-	}
-
-	// Start watching over the file and wait for events/errors.
-	if err := w.notifier.Add(w.file); err != nil {
-		if os.IsNotExist(err) {
-			// File got deleted in between initial check above and adding a watch.
-			w.isDeleted = true
-			w.updateTime = time.Now().UTC()
+		// If file wasn't deleted, no reason for notifier to have failed.
+		if notifierErr != nil {
```

Comment:
Note: here we could retry the notifier.Add instead of returning the error, but I don't think that's necessary. Agree?

Reply:
This is the case when the file didn't exist before. I don't think it will reach here no matter it is created (var returned), remains deleted (retry), or ctx is canceled (zeroVar returned). Do I miss anything? Is there a test case that shows what this check is used for?

Reply:
I agree that it shouldn't get here, and I can't easily write a test that does. However, it's possible that notifier.Add fails for some reason other than "file doesn't exist", and I think it's preferable to return that error here than to go into the select{}, which will hang forever.

Reply:
Yeah let's leave it here.